File: firmware/components/graph_engine/graph_engine.cpp
#include "graph_engine.h"
#include "calc_engine.h"

#include <cmath>

void graph_window_init(graph_window_t *w, double xmin, double xmax, double ymin, double ymax,
                       double xscl, double yscl) {
    if (!w) return;
    w->x_min = xmin;
    w->x_max = xmax;
    w->y_min = ymin;
    w->y_max = ymax;
    w->x_scl = xscl;
    w->y_scl = yscl;
}

void graph_math_to_pixel(const graph_window_t *w, double x, double y,
                         int width, int height, float *out_px, float *out_py) {
    if (!w || !out_px || !out_py) return;
    *out_px = (float)(((x - w->x_min) / (w->x_max - w->x_min)) * width);
    *out_py = (float)(height - ((y - w->y_min) / (w->y_max - w->y_min)) * height);
}

void graph_pixel_to_math(const graph_window_t *w, float px, float py,
                         int width, int height, double *out_x, double *out_y) {
    if (!w || !out_x || !out_y) return;
    *out_x = w->x_min + (px / (double)width) * (w->x_max - w->x_min);
    *out_y = w->y_min + ((height - py) / (double)height) * (w->y_max - w->y_min);
}

int graph_sample_y(const graph_window_t *w, const expr_handle_t *expr, angle_mode_t mode,
                   int width, int height, graph_sample_t *out, int max_samples) {
    if (!w || !expr || !out || max_samples <= 0) return 0;
    const int n = max_samples;
    for (int i = 0; i < n; ++i) {
        const double t = (n == 1) ? 0.0 : (double)i / (double)(n - 1);
        const double math_x = w->x_min + t * (w->x_max - w->x_min);
        const float px = (float)(t * (width - 1));
        const double math_y = calc_evaluate_at(expr, math_x, mode);
        out[i].pixel_x = px;
        out[i].math_y = (float)math_y;
        if (std::isfinite(math_y)) {
            float py;
            graph_math_to_pixel(w, math_x, math_y, width, height, &out[i].pixel_x, &py);
            out[i].pixel_x = px;
            out[i].pixel_y = py;
            out[i].valid = true;
        } else {
            out[i].pixel_y = NAN;
            out[i].valid = false;
        }
        (void)height;
    }
    return n;
}

File: GraphEngine.js
/**
 * GraphEngine.js
 * Coordinate transforms and curve sampling. DOM-free; depends on CalcEngine.evaluateAt
 * and global `math` (math.js) for expression compilation. ESP32 will replace math.js with TinyExpr.
 */

function resolveCalcEngine() {
  if (typeof CalcEngine !== 'undefined') return CalcEngine;
  if (typeof globalThis !== 'undefined' && globalThis.CalcEngine) return globalThis.CalcEngine;
  return require('./CalcEngine');
}

class GraphEngine {
  constructor(xMin = -10, xMax = 10, yMin = -10, yMax = 10, xScl = 1, yScl = 1) {
    this.xMin = xMin;
    this.xMax = xMax;
    this.yMin = yMin;
    this.yMax = yMax;
    this.xScl = xScl;
    this.yScl = yScl;
  }

  // Maps real math coordinates (x, y) to canvas pixel coordinates (pixelX, pixelY)
  mathToPixel(x, y, width, height) {
    const pixelX = ((x - this.xMin) / (this.xMax - this.xMin)) * width;
    const pixelY = height - ((y - this.yMin) / (this.yMax - this.yMin)) * height;
    return { x: pixelX, y: pixelY };
  }

  // Maps canvas pixel coordinates (pixelX, pixelY) back to real math coordinates (x, y)
  pixelToMath(pixelX, pixelY, width, height) {
    const x = this.xMin + (pixelX / width) * (this.xMax - this.xMin);
    const y = this.yMin + ((height - pixelY) / height) * (this.yMax - this.yMin);
    return { x, y };
  }

  /**
   * Coerce any complex, fraction, or bigNumber to a real number.
   * Imaginary residuals below 1e-12 are treated as real.
   */
  static coerceToNumber(val) {
    if (typeof val === 'number') {
      return val;
    }
    if (val && typeof val.toNumber === 'function') {
      return val.toNumber();
    }
    if (val && typeof val.im === 'number' && typeof val.re === 'number') {
      return Math.abs(val.im) < 1e-12 ? val.re : NaN;
    }
    return NaN;
  }

  coerceToNumber(val) {
    return GraphEngine.coerceToNumber(val);
  }

  /**
   * Generates a coordinate vector of pixel points for a given Y = f(x) equation.
   * Runs independently of the DOM.
   */
  generateYPoints(expression, width, height, angleMode = 'rad') {
    if (!expression || expression.trim() === '') {
      return { points: [], compiled: null, error: null };
    }

    const CE = resolveCalcEngine();

    try {
      const compiled = math.compile(expression);
      const points = [];

      for (let pixelX = 0; pixelX <= width; pixelX++) {
        const mathX = this.xMin + (pixelX / width) * (this.xMax - this.xMin);
        try {
          // evaluateAt already returns a coerced real number or NaN
          const mathY = CE.evaluateAt(compiled, mathX, angleMode);

          if (typeof mathY === 'number' && !isNaN(mathY) && isFinite(mathY)) {
            const pixelY = height - ((mathY - this.yMin) / (this.yMax - this.yMin)) * height;
            points.push({ x: pixelX, y: pixelY, mathY: mathY });
          } else {
            points.push({ x: pixelX, y: NaN, mathY: mathY });
          }
        } catch (evalErr) {
          points.push({ x: pixelX, y: NaN, mathY: NaN });
        }
      }

      return { points, compiled, error: null };
    } catch (parseErr) {
      return { points: [], compiled: null, error: parseErr.message };
    }
  }

  /**
   * Generates the pixel X coordinate for a vertical line X = c.
   */
  generateXPoint(expression, width, angleMode = 'rad') {
    if (!expression || expression.trim() === '') {
      return { pixelX: null, mathX: null, error: null };
    }

    const CE = resolveCalcEngine();

    try {
      const compiled = math.compile(expression);
      const val = CE.evaluateAt(compiled, 0, angleMode);
      if (typeof val !== 'number' || isNaN(val) || !isFinite(val)) {
        throw new Error("Expression did not evaluate to a constant real number.");
      }

      const pixelX = ((val - this.xMin) / (this.xMax - this.xMin)) * width;
      return { pixelX, mathX: val, error: null };
    } catch (parseErr) {
      return { pixelX: null, mathX: null, error: parseErr.message };
    }
  }
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = GraphEngine;
} else {
  window.GraphEngine = GraphEngine;
}
